File: 1043.partition-array-for-maximum-sum.ts
// @leet start
function maxSumAfterPartitioning(arr: number[], k: number): number {
  let cache = new Map();
  let dfs = (i: number): number => {
    if (cache.has(i)) return cache.get(i)!;
    let max = 0;
    let res = 0;
    for (let j = i; j < Math.min(i + k, arr.length); j++) {
      max = Math.max(max, arr[j]);
      let size = j - i + 1;
      res = Math.max(res, max * size + dfs(j + 1));
    }
    cache.set(i, res);
    return res;
  }

  return dfs(0);
};
// @leet end

module.exports = maxSumAfterPartitioning

// depth first search
// [1] -> [1] + [15] -> [1] + [15] + [7]
// [1] -> [1] + [15] -> [1] + [15] + [7] .. [5] [10] [0]
// [1] -> [1] + [15] -> [1] + [15] + [7] .. [10, 10] [0]

File: 1043.partition-array-for-maximum-sum.cs
// @leet start
public class Solution {
    public int MaxSumAfterPartitioning(int[] arr, int k) {
      var cache = new int[arr.Length + 1];
      Func<int, int> dfs = null;
      dfs = (i) => {
        if (cache[i] > 0) return cache[i];
        var max = 0;
        var res = 0;
        for (var j = i; j < Math.Min(i + k, arr.Length); j++) {
          max = Math.Max(max, arr[j]);
          res = Math.Max(res, max * (j - i + 1) + dfs(j + 1));
        }
        cache[i] = res;
        return res;
      };

      return dfs(0);
    }
}
// @leet end
